src/topology/inseparable.lean

```diff
@@ -178,12 +179,199 @@ def specialization_preorder : preorder X :=
+  (H : ∀ s ∈ S, stable_under_generalization s) : stable_under_generalization (⋂₀ S) :=
+is_upper_set_sInter H
+
+lemma stable_under_specialization_Union {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_specialization (S i)) : stable_under_specialization (⋃ i, S i) :=
+is_lower_set_Union H
+
+lemma stable_under_specialization_Inter {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_specialization (S i)) : stable_under_specialization (⋂ i, S i) :=
+is_lower_set_Inter H
+
+lemma stable_under_generalization_Union {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋃ i, S i) :=
+is_upper_set_Union H
+
+lemma stable_under_generalization_Inter {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋂ i, S i) :=
+is_upper_set_Inter H
+
+lemma stable_under_specialization_iff_eq_Union {s : set X} :
+  stable_under_specialization s ↔ s = ⋃ x ∈ s, closure {x} :=
+by simpa only [closure_singleton_eq_Iic] using is_lower_set_iff_eq_Inter_Iic
+
+alias stable_under_specialization_iff_eq_Union ↔ stable_under_specialization.eq_Union _
+
+/-- A set is stable under specialization iff it is a union of closed sets. -/
+lemma stable_under_specialization_iff_exists_sUnion_eq {s : set X} :
+  stable_under_specialization s ↔ ∃ (S : set (set X)), (∀ s ∈ S, is_closed s) ∧ ⋃₀ S = s :=
+begin
+  refine ⟨λ H, ⟨(λ x : X, closure {x}) '' s, _, _⟩, λ ⟨S, hS, e⟩, e ▸
+    stable_under_specialization_sUnion S (λ x hx, (hS x hx).stable_under_specialization)⟩,
+  { rintros _ ⟨_, _, rfl⟩, exact is_closed_closure },
+  { conv_rhs { rw stable_under_specialization_iff_eq_Union.mp H }, simp }
+end
+
+/-- A set is stable under generalization iff it is an intersection of open sets. -/
+lemma stable_under_generalization_iff_exists_sInter_eq {s : set X} :
+  stable_under_generalization s ↔ ∃ (S : set (set X)), (∀ s ∈ S, is_open s) ∧ ⋂₀ S = s :=
+begin
+  refine ⟨_, λ ⟨S, hS, e⟩, e ▸
+    stable_under_generalization_sInter S (λ x hx, (hS x hx).stable_under_generalization)⟩,
+  rw [← stable_under_specialization_compl_iff, stable_under_specialization_iff_exists_sUnion_eq],
+  exact λ ⟨S, h₁, h₂⟩, ⟨has_compl.compl '' S, λ s ⟨t, ht, e⟩, e ▸ (h₁ t ht).is_open_compl,
+    compl_injective ((sUnion_eq_compl_sInter_compl S).symm.trans h₂)⟩
+end
+
+lemma stable_under_specialization.preimage {s : set Y}
+  (hs : stable_under_specialization s) (hf : continuous f) :
+  stable_under_specialization (f ⁻¹' s) :=
+is_lower_set.preimage hs hf.specialization_monotone
+
+lemma stable_under_generalization.preimage {s : set Y}
+  (hs : stable_under_generalization s) (hf : continuous f) :
+  stable_under_generalization (f ⁻¹' s) :=
+is_upper_set.preimage hs hf.specialization_monotone
+
+/-- A map `f` between topological spaces is specializing if specializations lifts along `f`,
+i.e. for each `y ⬿ f x'` there is some `x ⬿ x'` whose image is `y`. -/
```

Comment:
Could you please switch to the "left-to-right" arrows here? We only use `⤳` as a `mathlib` notation, so I would use the same symbol in the docs.

Reply:
That would be slightly more verbose though: "for each `y` such that `f x' ⤳ y` there is some `x` such that `x' ⤳ x` whose image is `y`".

Reply:
I agree with urkud's request here. I don't think you even need to do the "such that" dance if it seems verbose, just write "for each f x' ⤳ y there is".